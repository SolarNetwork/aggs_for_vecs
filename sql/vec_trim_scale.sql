-- trim_scale all null
SELECT vec_trim_scale(ARRAY[NULL, NULL, NULL]::numeric[]);
-- trim_scale no trailing zeros
SELECT vec_trim_scale(ARRAY['1','2.2','3.03']::numeric[]);
-- trim_scale trailing zeros
SELECT vec_trim_scale(ARRAY[100,2.00,3.45600,400.0040]::numeric[]);
-- trim_scale mean lots
SELECT vec_trim_scale(vec_to_mean(nums)) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- trim_scale from measurements2 query
-- results expected to match:
-- WITH da AS (
--      SELECT idx_i, avg(val_a) AS avg_a
--      FROM measurements2 d, unnest(d.data_a) WITH ORDINALITY AS a(val_a, idx_i)
--      GROUP BY idx_i
-- )
-- SELECT array_agg(avg_a ORDER BY idx_i) AS data_a_avg FROM da
-- 
-- {6330897.179347826087,NULL,61.7650000000000000}
SELECT vec_trim_scale(vec_to_mean(pad_vec(data_a,3))) FROM measurements2;
