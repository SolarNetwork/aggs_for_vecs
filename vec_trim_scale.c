
Datum vec_trim_scale(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(vec_trim_scale);

// trim the output scale to drop trailing zeros; in PG13 the trim_scale function is available,
// otherwise manually find the appropriate scale to trim to
Datum
trimScaleNumeric(Datum num);

/**
 * Trims trailing zeros from a numeric array.
 *
 * This function takes an array of n numeric elements and performs a
 * "trim scale" operation on each to remove any trailing zeros by 
 * adjusting their decimal scale.
 */
Datum
vec_trim_scale(PG_FUNCTION_ARGS)
{
  int16 elemTypeWidth;
  bool elemTypeByValue;
  char elemTypeAlignmentCode;
  int valsLength;
  ArrayType *valsArray, *retArray;
  Datum *valsContent, *retContent;
  bool *valsNulls, *retNulls;
  int i;
  int dims[1];
  int lbs[1];

  if (PG_ARGISNULL(0)) {
    PG_RETURN_NULL();
  }

  valsArray = PG_GETARG_ARRAYTYPE_P(0);

  if (ARR_NDIM(valsArray) == 0) {
    PG_RETURN_NULL();
  }
  if (ARR_NDIM(valsArray) > 1) {
    ereport(ERROR, (errmsg("vec_trim_scale: one-dimensional arrays are required")));
  }

  get_typlenbyvalalign(NUMERICOID, &elemTypeWidth, &elemTypeByValue, &elemTypeAlignmentCode);
  deconstruct_array(valsArray, NUMERICOID, elemTypeWidth, elemTypeByValue, elemTypeAlignmentCode,
      &valsContent, &valsNulls, &valsLength);

  retContent = palloc0(sizeof(Datum) * valsLength);
  retNulls = palloc0(sizeof(bool) * valsLength);

  for (i = 0; i < valsLength; i++) {
    if (valsNulls[i]) {
      retNulls[i] = true;
      continue;
    }
    retNulls[i] = false;
    retContent[i] = trimScaleNumeric(valsContent[i]);
  }

  dims[0] = valsLength;
  lbs[0] = 1;

  retArray = construct_md_array(retContent, retNulls, 1, dims, lbs,
      NUMERICOID, elemTypeWidth, elemTypeByValue, elemTypeAlignmentCode);

  PG_RETURN_ARRAYTYPE_P(retArray);
}

Datum
trimScaleNumeric(Datum num) {
#if PG_VERSION_NUM >= 130000
   return DirectFunctionCall1(trim_scale, num);
#else
  char *str;
  int len;
  int j;
  int p;
  // look for number of trailing zeros in string version, then set scale accordingly
  str = DatumGetCString(DirectFunctionCall1(numeric_out, num));
  len = strlen(str);
  p = -1;
  for (j = len - 1; j >= 0; j--) {
    if (str[j] == '0') {
      continue;
    } else if (str[j] == '.') {
        // we found the decimal point; handle all trailing zeros by setting scale to 0
        return DirectFunctionCall2(numeric_trunc, num, Int32GetDatum(p > j ? p - j : 0));
    } else if (p < 0) {
      p = j;
    }
  }
  return num;
#endif
}
