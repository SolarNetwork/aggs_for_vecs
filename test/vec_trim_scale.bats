load test_helper

@test "trim_scale all null" {
  result="$(query "SELECT vec_trim_scale(ARRAY[NULL, NULL, NULL]::numeric[])")";
  echo $result;
  [ "$result" = "{NULL,NULL,NULL}" ]
}

@test "trim_scale no trailing zeros" {
  result="$(query "SELECT vec_trim_scale(ARRAY['1','2.2','3.03']::numeric[])")";
  echo $result;
  [ "$result" = "{1,2.2,3.03}" ]
}

@test "trim_scale trailing zeros" {
  result="$(query "SELECT vec_trim_scale(ARRAY[100,2.00,3.45600,400.0040]::numeric[])")";
  echo $result;
  [ "$result" = "{100,2,3.456,400.004}" ]
}

@test "trim_scale mean lots" {
  result="$(query "SELECT vec_trim_scale(vec_to_mean(nums)) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{1.23,2.34,2.895}" ]
}

# trim_scale from measurements2 query results calculated via:
# WITH da AS (
# 	SELECT idx_i, avg(val_a) AS avg_a
# 	FROM measurements2 d, unnest(d.data_a) WITH ORDINALITY AS a(val_a, idx_i)
# 	GROUP BY idx_i
# )
# SELECT array_agg(avg_a ORDER BY idx_i) AS data_a_avg FROM da
# 
# {6330897.179347826087,NULL,61.7650000000000000}

@test "trim_scale from measurements2 query" {
  result="$(query "SELECT vec_trim_scale(vec_to_mean(pad_vec(data_a,3))) FROM measurements2;")";
  echo $result;
  [ "$result" = "{6330897.179347826087,NULL,61.765}" ]
}
