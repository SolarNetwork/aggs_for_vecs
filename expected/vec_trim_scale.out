-- trim_scale all null
SELECT vec_trim_scale(ARRAY[NULL, NULL, NULL]::numeric[]);
  vec_trim_scale  
------------------
 {NULL,NULL,NULL}
(1 row)

-- trim_scale no trailing zeros
SELECT vec_trim_scale(ARRAY['1','2.2','3.03']::numeric[]);
 vec_trim_scale 
----------------
 {1,2.2,3.03}
(1 row)

-- trim_scale trailing zeros
SELECT vec_trim_scale(ARRAY[100,2.00,3.45600,400.0040]::numeric[]);
    vec_trim_scale     
-----------------------
 {100,2,3.456,400.004}
(1 row)

-- trim_scale mean lots
SELECT vec_trim_scale(vec_to_mean(nums)) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
  vec_trim_scale   
-------------------
 {1.23,2.34,2.895}
(1 row)

-- trim_scale from measurements2 query
-- results expected to match:
-- WITH da AS (
--      SELECT idx_i, avg(val_a) AS avg_a
--      FROM measurements2 d, unnest(d.data_a) WITH ORDINALITY AS a(val_a, idx_i)
--      GROUP BY idx_i
-- )
-- SELECT array_agg(avg_a ORDER BY idx_i) AS data_a_avg FROM da
-- 
-- {6330897.179347826087,NULL,61.7650000000000000}
SELECT vec_trim_scale(vec_to_mean(pad_vec(data_a,3))) FROM measurements2;
           vec_trim_scale           
------------------------------------
 {6330897.179347826087,NULL,61.765}
(1 row)

